components/pkg/utils/snowflake/ringbuffer.go

```diff
@@ -0,0 +1,186 @@
+
+	//asynchronously running padding goroutine numbers
+	GoNum int32
+}
+
+var cores int = runtime.NumCPU()
+
+func NewRingBuffer(bufferSize int64) *RingBuffer {
+	p := PaddedInt{}
+	p.Value = -1
+	return &RingBuffer{
+		slots:      make([]int64, bufferSize),
+		flags:      make([]PaddedInt, bufferSize),
+		wp:         p,
+		rp:         p,
+		bufferSize: bufferSize,
+		GoNum:      1,
+	}
+}
+
+func (r *RingBuffer) Put(uid int64) (bool, error) {
+	r.m.Lock()
+	defer r.m.Unlock()
+	currentWritePointer := r.wp.Value
+	currentReadPointer := r.rp.Value
+	if currentReadPointer == -1 {
+		currentReadPointer = 0
+	}
+	distance := currentWritePointer - currentReadPointer
+	//write pointer catches read pointer, ringbuffer is full
+	if distance == r.bufferSize-1 {
+		return false, errors.New("ringbuffer is full! Rejected putting buffer")
+	}
+
+	//(currentWritePointer + 1) mod r.bufferSize
+	nextWriteIndex := (currentWritePointer + 1) & (r.bufferSize - 1)
+
+	if r.flags[nextWriteIndex].Value != CAN_PUT_FLAG {
+		return false, errors.New("slot is not in writable status")
+	}
+
+	r.slots[nextWriteIndex] = uid
+	r.flags[nextWriteIndex].Value = CAN_TAKE_FLAG
+	r.wp.Value++
+	return true, nil
+}
+
+func (r *RingBuffer) Take() (int64, error) {
+	r.m.Lock()
+	defer r.m.Unlock()
+	var uid int64
+
+	if r.rp.Value != r.wp.Value {
+		r.rp.Value++
+	} else {
+		return uid, errors.New("buffer is empty, rejected take buffer")
+	}
+
+	if r.wp.Value-r.rp.Value < r.bufferSize*r.PaddingFactor/100 {
+		//limit the numbers of goroutine
+		if int(r.GoNum) <= 2*cores {
+			r.GoNum++
+			go r.PaddingRingBuffer()
+		}
+	}
+
+	//check next slot flag is CAN_TAKE_FLAG
+	nextReadIndex := (r.rp.Value) & (r.bufferSize - 1)
+	if r.flags[nextReadIndex].Value != CAN_TAKE_FLAG {
+		return uid, errors.New("slot not in readable status")
+	}
+
+	uid = r.slots[nextReadIndex]
+	r.flags[nextReadIndex].Value = CAN_PUT_FLAG
+	return uid, nil
+}
+
+//allocate bits for uid
+//workid + time + sequence
+func (r *RingBuffer) Allocator() int64 {
+	var sequence int64
+
+	timestampShift := r.SeqBits
+	workidShift := r.TimeBits + r.SeqBits
+
+	workid := r.WorkId
+
+	r.m.Lock()
+	timestamp := r.CurrentTimeStamp
+	r.CurrentTimeStamp++
+	r.m.Unlock()
+
+	return timestamp<<timestampShift | (workid << workidShift) | sequence
+}
+
+//put uid into ringbuffer
+//uid： workid + timestamp + (0 ~ maxSeq)
+func (r *RingBuffer) GenerateUid(cur int64) (bool, error) {
+	var maxSeq int64 = ^(-1 << r.SeqBits) + 1
+	var offset int64
+	for offset = 0; offset < maxSeq; offset++ {
+		if ok, err := r.Put(cur + offset); !ok {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+//async padding ringbuffer
+func (r *RingBuffer) PaddingRingBuffer() {
+	defer func() {
+		if x := recover(); x != nil {
+			log.DefaultLogger.Errorf("panic when generatoring uid with snowflake algorithm and padding ringbuffer: %v", x)
+		}
+	}()
+
+	for {
+		u := r.Allocator()
+		if ok, err := r.GenerateUid(u); !ok || err != nil {
```

Comment:
`Allocator ` func allocate `workid + time` and then` GenerateUid ` allocate offset？

Reply:
yes